### PR TITLE
[bare-code-scanner] Fix only scans on first mount on Android

### LINKED ISSUE #9619: Once Barcode is Scanned, Cant Scan again
SO I have a scanner to scan a barcode and extract the numbers from it. if you do it the first time it works. after the barcode is submitting and you push the camera button I made to open the modal that scanner is attached to, it will not let you scan it again.

### LINKED ISSUE #15145: BarCodeScanner only scans on first mount Android
### Summary

Responsive to upgrade from Expo SDK 42 => 43 on Android (only).  expo-barcode-scanner will only scan once.  After that, the camera will still come up but it will not detect any bar code.  To scan again you must not only close the application but also Expo Go and restart.  Then you will get one scan only before doing another restart.  

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

43

### Environment

 Expo CLI 4.13.0 environment info:
    System:
      OS: Windows 10 10.0.19042
    Binaries:
      Node: 14.17.3 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 7.20.5 - C:\Program Files\nodejs\npm.CMD
    npmPackages:
      expo: ^43.0.2 => 43.0.2
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
    Expo Workflow: managed

### Reproducible demo or steps to reproduce from a blank project

This can be reproduced in the Basic BarCodeScanner usage snack from the Expo website API reference: exp://exp.host/@snack/sdk.43.0.0-fiKiGK2vqE

1. Scan using the snack once.
2. Close the snack keeping Expo Go open.
3. Reload the snack from Expo Go.  Notice it will not find a barcode.
4. Close the snack AND Expo Go. 
5. Start Expo Go again and load the snack.  You will can again (just once). 

In other words, it requires relaunching the Expo Go.  This is the same behavior our own applications.

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Fixes https://github.com/expo/expo/issues/15145.
Fixes https://github.com/expo/expo/issues/9619.
Closes ENG-2490.

# How

`barCodeScannerTaskLock` wasn't set to false when the coroutine scope was canceled, because if your try to run something in the context that was canceled, that instruction will be ignored. That's why we need to check if the scope is active before running `launch`. 

# Test Plan

- expo-go using https://snack.expo.dev/@ajhopen/1x-barcode-scan-issue-sdk43 ✅

### CODE REVIEW

**Review by tsapeta — APPROVED** [review]

**Comment by Fabzien** [comment]
Hi,

I just upgraded to SDK 44 and I still have the same behavior. Only first first mount is working. I don't have any issue on iOS build.

My dependencies:

> "dependencies": {
>     "@react-native-async-storage/async-storage": "~1.15.0",
>     "@react-native-community/masked-view": "0.1.10",
>     "@react-native-picker/picker": "^2.2.1",
>     "expo": "^43.0.0",
>     "expo-barcode-scanner": "~11.1.2",
>     "expo-camera": "~12.0.3",
>     "expo-file-system": "~13.0.3",
>     "expo-localization": "~11.0.0",
>     "expo-mail-composer": "~11.0.3",
>     "expo-permissions": "~13.0.3",
>     "react": "17.0.1",
>     "react-dom": "17.0.1",
>     "react-native": "0.64.3",
>     "react-native-collapsible": "^1.6.0",
>     "react-native-gesture-handler": "~1.10.2",
>     "react-native-i18n": "^2.0.15",
>     "react-native-reanimated": "^1.13.3",
>     "react-native-safe-area-context": "3.3.2",
>     "react-native-screens": "~3.8.0",
>     "react-native-table-component": "^1.2.1",
>     "react-native-web": "0.17.1",
>     "react-navigation": "^4.2.2",
>     "react-navigation-drawer": "^2.4.2",
>     "react-navigation-stack": "^2.2.3"
>   },

Any ideas ?

**Comment by bbarthec** [comment]
@Fabzien, you're saying that you've upgraded to SDK 44, but I can see that these dependencies' versions are related to SDK 43 🤔 As you can read in [`expo-barcode-scanner/CHANGELOG file`](https://github.com/expo/expo/blob/master/packages/expo-barcode-scanner/CHANGELOG.md) the fix you're looking for is available in `expo-barcode-scanner@11.2.0` (and you're having version `11.1.2`).
Try to actually upgrade to SDK 44 and check again 👀

**Comment by Fabzien** [comment]
@bbarthec 

My bad, I must have copied my old package.json file.

Here after upgrade & expo start --clear, still same issue:

>   "dependencies": {
    "@react-native-async-storage/async-storage": "~1.15.0",
    "@react-native-community/masked-view": "0.1.10",
    "@react-native-picker/picker": "2.2.1",
    "expo": "^44.0.0",
    "expo-barcode-scanner": "~11.2.0",
    "expo-camera": "~12.1.0",
    "expo-file-system": "~13.1.0",
    "expo-localization": "~12.0.0",
    "expo-mail-composer": "~11.1.0",
    "expo-permissions": "~13.1.0",
    "react": "17.0.1",
    "react-dom": "17.0.1",
    "react-native": "0.64.3",
    "react-native-collapsible": "^1.6.0",
    "react-native-gesture-handler": "~2.1.0",
    "react-native-i18n": "^2.0.15",
    "react-native-reanimated": "~2.3.1",
    "react-native-safe-area-context": "3.3.2",
    "react-native-screens": "~3.10.1",
    "react-native-table-component": "^1.2.1",
    "react-native-web": "0.17.1",
    "react-navigation": "^4.2.2",
    "react-navigation-drawer": "^2.4.2",
    "react-navigation-stack": "^2.2.3"
  },

**Comment by sk1llsh0t** [comment]
i upgraded from sdk 43 to 44.0.1 and barcode scanner 11.2.0. everything appears to be working properly now.

**Comment by Fabzien** [comment]
@sk1llsh0t @bbarthec 
Thank you for answer. I made some progress with SDK 44 !

Case 1:

- Scanning a QR & display my result screen
- Click back button
- Trying scanning another QR = Not detected

Case 2:

- Scanning a QR & display my result screen
- Click back button (so back to my "camera screen") then back again to my main screen
- Click on "Scan QR" button 
- Scanning a QR = Detected then display my result screen

In conclusion, QR code scanning is not working directly after using back button on my result screen. Maybe "barcode scanner" is not mount or any other idea ?

On iOS everything is fine in both cases I can scan a QR code.

**Comment by sk1llsh0t** [comment]
it could be your use case that might be causing the issue. 

1) i don't use qr codes. i use 1D barcodes when i scan so maybe there is an issue there
2) from my result screen, i tap a scan button to open the scan screen. after scanning or tapping cancel, it takes me back to the result screen. maybe it is the difference in the order in which the screens appear. since your scan screen is always  in the stack, maybe it isn't getting reloaded correctly in that case. in my case, the scan screen "unloads" every time it is used.

**Comment by bbarthec** [comment]
@Fabzien, given you've describing two cases, I'd like you to create an issue containing an MVP example that would present the whole scenario and this buggy situation you're experiencing - using snack for demo app would be the best 😊 
This is the quickest way for us to fix the problem, because we'll be having a living example that we would base on while debugging the problem 😉

**Comment by Fabzien** [comment]
@bbarthec 

Here a working example:

https://snack.expo.dev/@fabzien/bare-code-scanner-report-issue

I had to setup project with expo v43 because v44 is not available on Snack.

To reproduce (Android only):

1. Click "Go to QRCodeScanner"
2. Scan any QR Code
3. Click "Back" on next screen
4. Click "Tap to Scan Again"
5. Try to scan another QR code = Failed

Do you want me to open a regular issue ?

**Comment by bbarthec** [comment]
@Fabzien, that would be best 👍 
You can add a reference to this thread/discussion in this newly created issues as well 😊

**Comment by kchangch** [comment]
Currently I have this:

` "expo": "^44.0.1",
    "expo-barcode-scanner": "^11.2.0",`

But it seems that my android barcode scanner only scan once. Any help?

**Comment by andrconstruction** [comment]
the same situation "expo": "^45.0.5", "expo-barcode-scanner": "^11.3.0", scan only once

**Comment by AndonMitev** [comment]
same situation with `"expo-barcode-scanner": "11.4.0", "expo-barcode-scanner": "11.4.0"` after scan and press `Scan Again` component is not rendering
